src/Microsoft.AspNetCore.Routing.Abstractions/MetadataCollection.cs

```diff
@@ -0,0 +1,101 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public class EndpointMetadataCollection : IReadOnlyList<object>
+    {
+        public static readonly EndpointMetadataCollection Empty = new EndpointMetadataCollection(Array.Empty<object>());
+
+        private readonly object[] _items;
+
+        public EndpointMetadataCollection(IEnumerable<object> items)
+        {
+            if (items == null)
+            {
+                throw new ArgumentNullException(nameof(items));
+            }
+
+            _items = items.ToArray();
+        }
+
+        public object this[int index] => _items[index];
+
+        public int Count => _items.Length;
+
+        public T GetMetadata<T>() where T : class
+        {
+            for (var i = _items.Length -1; i >= 0; i--)
+            {
+                var item = _items[i] as T;
+                if (item !=null)
+                {
+                    return item;
+                }
+            }
+
+            return default;
+        }
+
+        public IEnumerable<T> GetOrderedMetadata<T>() where T : class
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                var item = _items[i] as T;
+                if (item != null)
+                {
+                    yield return item;
+                }
+            }
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(this);
+
+        IEnumerator<object> IEnumerable<object>.GetEnumerator() => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public struct Enumerator : IEnumerator<object>
+        {
+            private object[] _items;
```

Comment:
This is intentionally not-readonly to prevent defensive copies

Reply:
I'm not sure that there's actually a problem here, it's just my habit to make private fields mutable on structs to avoid that, I don't think it matters much and if it does we'll find it later.